hyperspy/events.py

```diff
@@ -0,0 +1,508 @@
+            new_doc += '\t%s :\n\t\t%s\n' % (name, e_short)
+        new_doc = new_doc.replace('\t', '    ')
+        self.__doc__ = new_doc
+
+    def __setattr__(self, name, value):
+        """
+        Magic to enable having `Event`s as attributes, and keeping them
+        separate from other attributes.
+
+        If it's an `Event`, store it in self._events, otherwise set attribute
+        in normal way.
+        """
+        if isinstance(value, Event):
+            self._events[name] = value
+            self._update_doc()
+        else:
+            super(Events, self).__setattr__(name, value)
+
+    def __getattr__(self, name):
+        """
+        Magic to enable having `Event`s as attributes, and keeping them
+        separate from other attributes.
+
+        Returns Event attribute `name` (__getattr__ is only called if attribute
+        could not be found in the normal way).
+        """
+        return self._events[name]
+
+    def __delattr__(self, name):
+        """
+        Magic to enable having `Event`s as attributes, and keeping them
+        separate from other attributes.
+
+        Deletes attribute from self._events if present, otherwise delete
+        attribute in normal way.
+        """
+        if name in self._events:
+            del self._events[name]
+            self._update_doc()
+        else:
+            super(Events, self).__delattr__(name)
+
+    def __dir__(self):
+        """
+        Magic to enable having `Event`s as attributes, and keeping them
+        separate from other attributes.
+
+        Makes sure tab-completion works in IPython etc.
+        """
+        d = dir(type(self))
+        d.extend(self.__dict__.iterkeys())
+        d.extend(self._events.iterkeys())
+        return sorted(set(d))
+
+    def __iter__(self):
+        """
+        Allows iteration of all events in the container
+        """
+        return self._events.itervalues()
+
+    def __repr__(self):
+        text = "<hyperspy.events.Events: " + repr(self._events) + ">"
+        return text.encode('utf8')
+
+
+class Event(object):
+
+    def __init__(self, doc='', arguments=None):
+        """
+        Create an Event object.
+
+        Arguments:
+        ----------
+            doc : str
+                Optional docstring for the new Event.
+            arguments : iterable
+                Pass to define the arguments of the trigger() function. Each
+                element must either be an argument name, or a tuple containing
+                the argument name and the argument's default value.
+
+        Example usage:
+        --------------
+            >>> Event()
+            <hyperspy.events.Event: {}
+            >>> Event(doc="This event has a docstring!").__doc__
+            'This event has a docstring!'
+            >>> e1 = Event()
+            >>> e2 = Event(arguments=('arg1', ('arg2', None)))
+            >>> e1.trigger(12, 43, 'str', 4.3)  # Can trigger with whatever
```

Comment:
This line needs to be updated to only have keyword arguments, as it would currently fail.